File: libs/java/common/src/main/java/arda/common/model/ApiResponse.java
package arda.common.model;

import arda.common.exception.ErrorCode;
import java.time.Instant;
import java.util.List;
import java.util.Map;

public record ApiResponse<T>(
    boolean success,
    T data,
    ApiError error,
    Map<String, Object> metadata,
    Instant timestamp
) {
    public ApiResponse {
        if (timestamp == null) timestamp = Instant.now();
    }

    public static <T> ApiResponse<T> success(T data) {
        return new ApiResponse<>(true, data, null, null, Instant.now());
    }

    public static <T> ApiResponse<T> success(T data, Map<String, Object> metadata) {
        return new ApiResponse<>(true, data, null, metadata, Instant.now());
    }

    public static <T> ApiResponse<T> error(ErrorCode errorCode) {
        return new ApiResponse<>(false, null, new ApiError(errorCode.getCode(), errorCode.getMessage(), null), null, Instant.now());
    }

    public static <T> ApiResponse<T> error(ErrorCode errorCode, String message) {
        return new ApiResponse<>(false, null, new ApiError(errorCode.getCode(), message, null), null, Instant.now());
    }

    public record ApiError(String code, String message, List<ErrorDetail> details) {}

    public record ErrorDetail(String field, String reason, Map<String, Object> metadata) {}
}

File: arda-be-java/libs/common/src/main/kotlin/arda/common/model/ApiResponse.kt
package arda.common.model

import arda.common.exception.ErrorCode
import java.time.Instant

data class ApiResponse<T>(
    val success: Boolean,
    val data: T? = null,
    val error: ApiError? = null,
    val metadata: Map<String, Any>? = null,
    val timestamp: Instant = Instant.now()
) {
    companion object {
        fun <T> success(data: T?, metadata: Map<String, Any>? = null): ApiResponse<T> = ApiResponse(
            success = true,
            data = data,
            metadata = metadata
        )

        fun <T> error(errorCode: ErrorCode, message: String? = null, details: List<ErrorDetail>? = null): ApiResponse<T> = ApiResponse(
            success = false,
            error = ApiError(
                code = errorCode.code,
                message = message ?: errorCode.message,
                details = details
            )
        )
    }
}

data class ApiError(
    val code: String,
    val message: String,
    val details: List<ErrorDetail>? = null
)

data class ErrorDetail(
    val field: String? = null,
    val reason: String,
    val metadata: Map<String, Any>? = null
)
